File: portal/server/limits.js
/**
 * Per-user usage-limit defaults + resolution (interim auth backend).
 *
 * Two kinds of limit, both now per-user-configurable on top of a shared
 * "standard plan" default:
 *  - dailyTokenLimit   — server-enforced token ceiling per user per IST day.
 *  - contextSoft/Hard  — the per-conversation guardrail thresholds the SPA uses
 *                        to warn (soft) and hard-stop (hard). The hard limit can
 *                        be LOWERED per user but never raised past the model's
 *                        real 200k context window — the model rejects more.
 *
 * A user document carries an optional sparse `limits` object; any absent field
 * falls back to the default, so a freshly seeded user (no `limits`) gets the
 * standard plan and an admin raises a single field to "approve a higher plan".
 *
 * This module is the single source of truth, imported by the /api/claude gate,
 * the usage endpoint, the login/refresh profile, and the admin routes — so the
 * number a user is enforced against, billed against, and shown are all one
 * definition.
 */

// The Opus 4.7 context window. A per-user hard limit is clamped to this — you
// cannot grant more context than the model can accept.
export const MODEL_CONTEXT_WINDOW = 200_000

// The standard plan (defaults applied to every user without an override).
export const DEFAULT_DAILY_TOKEN_LIMIT = 1_000_000
export const DEFAULT_CONTEXT_SOFT_LIMIT = 150_000
export const DEFAULT_CONTEXT_HARD_LIMIT = 200_000

// The only fields an admin can override per user.
export const LIMIT_FIELDS = ['dailyTokenLimit', 'contextSoftLimit', 'contextHardLimit']

function posIntOr(value, fallback) {
  return Number.isInteger(value) && value > 0 ? value : fallback
}

/**
 * The standard-plan limits. Daily comes from DAILY_TOKEN_LIMIT (validated at
 * boot by server.js; a bad value falls back here too); context is fixed to the
 * constants. Read fresh each call so an env change between requests is honoured.
 */
export function defaultLimits() {
  const envDaily = Number.parseInt(process.env.DAILY_TOKEN_LIMIT, 10)
  return {
    dailyTokenLimit: envDaily > 0 ? envDaily : DEFAULT_DAILY_TOKEN_LIMIT,
    contextSoftLimit: DEFAULT_CONTEXT_SOFT_LIMIT,
    contextHardLimit: DEFAULT_CONTEXT_HARD_LIMIT,
  }
}

/**
 * Resolve a user's EFFECTIVE limits: per-user overrides on top of `defaults`,
 * each validated and clamped. Never throws on user/DB data — an invalid or
 * absent override silently falls back to the default. This is the final safety
 * net, so a partial admin patch that briefly stored soft >= hard still reads
 * back as a sane soft < hard <= window.
 *
 * @param {object|null} user - a user document (may carry `limits`), or null.
 */
export function resolveUserLimits(user, defaults = defaultLimits()) {
  const o = (user && typeof user === 'object' && user.limits) || {}
  const dailyTokenLimit = posIntOr(o.dailyTokenLimit, defaults.dailyTokenLimit)
  // Hard limit: per-user value (or default), never above the model window.
  const contextHardLimit = Math.min(
    posIntOr(o.contextHardLimit, defaults.contextHardLimit),
    MODEL_CONTEXT_WINDOW,
  )
  // Soft (warn) must stay strictly below hard, and at least 1.
  let contextSoftLimit = posIntOr(o.contextSoftLimit, defaults.contextSoftLimit)
  contextSoftLimit = Math.max(1, Math.min(contextSoftLimit, contextHardLimit - 1))
  return { dailyTokenLimit, contextSoftLimit, contextHardLimit }
}

/**
 * Validate an admin PATCH body for a user's limit overrides. Each field is
 * optional; a value of `null` clears that override (revert to default); a
 * number must be a positive integer. The admin UI sends soft+hard together, so
 * the cross-field `soft < hard` check applies when both are present; a single
 * field is still accepted (resolveUserLimits re-clamps on read).
 *
 * @returns {{ok:true, limits:object} | {ok:false, error:string}}
 */
export function validateLimitsPatch(body) {
  if (!body || typeof body !== 'object' || Array.isArray(body)) {
    return { ok: false, error: 'Request body must be an object of limit fields.' }
  }
  const limits = {}
  for (const field of LIMIT_FIELDS) {
    if (!(field in body)) continue
    const v = body[field]
    if (v === null) {
      limits[field] = null // clear → fall back to default
      continue
    }
    if (!Number.isInteger(v) || v <= 0) {
      return { ok: false, error: `${field} must be a positive integer, or null to reset to default.` }
    }
    limits[field] = v
  }
  if (Object.keys(limits).length === 0) {
    return { ok: false, error: 'Provide at least one of: dailyTokenLimit, contextSoftLimit, contextHardLimit.' }
  }
  if (Number.isInteger(limits.contextHardLimit) && limits.contextHardLimit > MODEL_CONTEXT_WINDOW) {
    return {
      ok: false,
      error: `contextHardLimit cannot exceed the model context window of ${MODEL_CONTEXT_WINDOW.toLocaleString('en-US')} tokens.`,
    }
  }
  if (
    Number.isInteger(limits.contextSoftLimit) &&
    Number.isInteger(limits.contextHardLimit) &&
    limits.contextSoftLimit >= limits.contextHardLimit
  ) {
    return { ok: false, error: 'contextSoftLimit must be less than contextHardLimit.' }
  }
  return { ok: true, limits }
}

File: backend/src/services/usage/limits.py
"""Effective per-user limits — the daily token cap and the per-conversation context
guardrails (soft/hard), resolved with the clamps the portal has always applied.

The SINGLE source of truth shared by the admin `/admin/users` endpoint AND `/auth/me`, so a
superadmin's per-user override reaches the client (the daily badge + the "getting long"
warning) instead of the client silently falling back to the global defaults. Daily reuses the
gate's resolver so the badge and the 429 gate can never diverge.

WHO ENFORCES WHICH. `hard` is the SERVER's: `usage/context_window.enforce_context_limit`
refuses a turn at the route, before anything is persisted. `soft` is the browser's: it is
advisory, it blocks nothing, and it exists to warn the citizen in time to start a new chat
rather than to be told at the wall. This docstring used to call both of them "the values the
client should enforce", which was true of neither — nothing enforced them at all, front or
back, while an administrator was being shown a field promising a hard stop.
"""

from __future__ import annotations

import uuid
from typing import Final

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from src.db.models.user_limit import UserLimit
from src.services.usage.gate import effective_daily_limit

# Context guardrails (Express `limits.js`) — the SPA per-conversation warn/stop thresholds.
MODEL_CONTEXT_WINDOW = 200_000
DEFAULT_CONTEXT_SOFT = 150_000
DEFAULT_CONTEXT_HARD = 200_000

SYSTEM_PROMPT_RESERVE: Final = 8_000
"""What a run costs before the citizen has typed anything — the per-run system prompt and the
tool schemas that ride with it.

Measured at the time of writing: the Plan segment composes to ~1,800 tokens and the Build
segment — the larger — to ~4,400, before the tool schemas. 8,000 covers the larger of the two
with room for the schemas and for both to grow.

★ NOTHING HOLDS IT BACK ANY MORE, AND THAT IS AN IMPROVEMENT RATHER THAN A LOSS. This used to
be a reserve in the literal sense: the gate estimated a conversation and then added this,
because the system prompt was composed after the gate had decided and the estimate could not
see it. The gate now reads the token count the PROVIDER reported for a completed turn, and that
count is of the whole prompt — system segment, tool schemas and all. There is nothing left for
it to be blind to, so there is nothing to reserve.

IT SURVIVES AS A SIZE, NOT AS A CHARGE. The floor below is derived from it, and the admin
panel's `contextLimits.SYSTEM_PROMPT_RESERVE` is its browser twin, so an administrator's lowest
settable ceiling is expressed in the same unit on both sides."""

CONTEXT_HARD_FLOOR: Final = SYSTEM_PROMPT_RESERVE * 2
"""The lowest per-user chat length that still leaves a usable chat.

AN ADMINISTRATOR MUST NOT BE ABLE TO LOCK A CITIZEN OUT, and below this they could. Every run
spends `SYSTEM_PROMPT_RESERVE` on its system prompt and tool schemas before the citizen has
typed a word, and the provider counts that in the very first turn it reports — so a hard limit
at or under the reserve refuses that person's SECOND message in every chat they open, forever,
whatever they wrote in the first. Only another administrator raising the number gets them
working again, and nothing in the product says that is what happened.

TWICE THE RESERVE, not the reserve plus one: a floor that merely clears the reserve would leave
a chat with room for a sentence and no reply. This leaves the reserve plus an equal amount of
real conversation — tight, deliberate, and still a chat someone can use."""


def effective_context(override: UserLimit | None) -> tuple[int, int]:
    """Resolve (soft, hard) with Express's clamps: hard ≤ the model window and ≥ the floor;
    soft in [1, hard-1]. A non-positive/absent override falls back to the default (0/negative
    never caps to nothing).

    THE FLOOR IS APPLIED AT READ TIME AS WELL AS AT WRITE TIME, and both halves are needed. The
    admin PATCH validator refuses a new value below it, which is where an administrator learns
    why; this clamp is what keeps a value ALREADY stored below the floor — written before the
    validator existed — from locking that citizen out of every chat they own. Validation alone
    would leave the people the defect already reached exactly where it left them."""
    hard_raw = (
        override.context_hard_limit
        if override and override.context_hard_limit and override.context_hard_limit > 0
        else DEFAULT_CONTEXT_HARD
    )
    hard = max(CONTEXT_HARD_FLOOR, min(hard_raw, MODEL_CONTEXT_WINDOW))
    soft_raw = (
        override.context_soft_limit
        if override and override.context_soft_limit and override.context_soft_limit > 0
        else DEFAULT_CONTEXT_SOFT
    )
    soft = max(1, min(soft_raw, hard - 1))
    return soft, hard


async def effective_limits_for(db: AsyncSession, user_id: uuid.UUID) -> tuple[int, int, int]:
    """(daily, context_soft, context_hard) for one user, as the client is told them: `daily`
    drives the badge, `context_soft` the browser's warning, `context_hard` the number the
    SERVER refuses at (sent so the browser can describe the boundary, never so it can police
    it). Loads the user's override once; daily via the gate resolver so the badge and the gate
    agree."""
    daily = await effective_daily_limit(db, user_id)
    override = await db.scalar(select(UserLimit).where(UserLimit.user_id == user_id))
    soft, hard = effective_context(override)
    return daily, soft, hard
